- [Main output](#main-output)
- [Modules](#modules)
  - [fastqc(optional)](#fastqcoptional)
  - [extract](#extract)
  - [bwa\_mem](#bwa_mem)
  - [make\_fragment](#make_fragment)
  - [snapatac2](#snapatac2)
  - [multiqc-sgr](#multiqc-sgr)
  - [pipeline\_info](#pipeline_info)


# Main output

- `multiqc/multiqc_report.html` HTML report containing QC metrics.
- `snapatac2/{sample}.h5ad` AnnData object of fragments, QC metrics and clusters.
- `snapatac2/{sample}.gene_mat.h5ad` AnnData object of cell by gene activity matrix
- `snapatac2/{sample}.peak_mat.h5ad` AnnData object of cell by peak matrix

For details on how to use h5ad files, please refer to
- [SnapATAC2: A Python/Rust package for single-cell epigenomics analysis](https://kzhang.org/SnapATAC2/index.html)
- [features, gene activity, promoter and peaks](https://github.com/kaizhang/SnapATAC2/issues/164)

# Modules


## fastqc(optional)

[FastQC](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/) gives general quality metrics about your sequenced reads. It provides information about the quality score distribution across your reads, per base sequence content (%A/T/G/C), adapter contamination and overrepresented sequences. For further reading and documentation see the [FastQC help pages](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/).

**Output files**

- `*_fastqc.html`: FastQC report containing quality metrics.
- `*_fastqc.zip`: Zip archive containing the FastQC report, tab-delimited data file and plot images.

## extract

Extract the barcode from R2 fastqs and add the barcode to the read names of R1 and R3 fastqs.

**Output files** 
- `{sample}.{R1,R3}.fq.gz` Fastq files with barcodes added.

## bwa_mem

Align R1 and R3 reads to the genome.

**Output files** 
- `{sample}.bam` Coordinate sorted BAM.

## make_fragment

Convert a BAM file to a fragment file using [snapatac2.pp.make_fragment_file](https://kzhang.org/SnapATAC2/version/2.6/api/_autosummary/snapatac2.pp.make_fragment_file.html)

**Output files** 
- `{sample}.fragment.gz` Fragment file.

## [snapatac2](https://kzhang.org/SnapATAC2/tutorials/index.html)

Preprocessing, Dimenstion reduction, Clustering analysis and Peak calling at the cluster-level.

## multiqc-sgr

[MultiQC](http://multiqc.info) is a visualization tool that generates a single HTML report summarising all samples in your project. Most of the pipeline QC results are visualised in the report and further statistics are available in the report data directory.

[multiqc-sgr](https://pypi.org/project/multiqc-sgr/) adds some modules on this basis to facilitate the visualization of single cell-related data.

**Output files**

- `multiqc_report.html`: a standalone HTML file that can be viewed in your web browser.
- `multiqc_data/`: directory containing parsed statistics from the different tools used in the pipeline.
- `multiqc_plots/`: directory containing static images from the report in various formats.

## pipeline_info

[Nextflow](https://www.nextflow.io/docs/latest/tracing.html) provides excellent functionality for generating various reports relevant to the running and execution of the pipeline. This will allow you to troubleshoot errors with the running of the pipeline, and also provide you with other information such as launch commands, run times and resource usage.

**Output files**

- Reports generated by Nextflow: `execution_report.html`, `execution_timeline.html`, `execution_trace.txt` and `pipeline_dag.dot`/`pipeline_dag.svg`.
- Reports generated by the pipeline: `pipeline_report.html`, `pipeline_report.txt` and `software_versions.yml`. The `pipeline_report*` files will only be present if the `--email` / `--email_on_fail` parameter's are used when running the pipeline.
- Reformatted samplesheet files used as input to the pipeline: `samplesheet.valid.csv`.
- Parameters used by the pipeline run: `params.json`.
